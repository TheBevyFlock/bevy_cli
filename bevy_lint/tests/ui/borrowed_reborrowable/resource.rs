//! This tests the `borrowed_reborrowable` lint, specifically when triggered on the `ResMut` type.

#![feature(register_tool)]
#![register_tool(bevy)]
#![deny(bevy::borrowed_reborrowable)]

use bevy::prelude::*;

#[derive(Resource)]
struct Data(String);

// OK: Lint does not apply to immutable references
fn immutable_reference(_res: &ResMut<Data>) {
    // ...
}

//~| HELP: use `ResMut` instead
//~v ERROR: parameter takes `&mut ResMut` instead of a re-borrowed `ResMut`
fn mutable_reference(_res: &mut ResMut<Data>) {
    // ...
}

//~| HELP: use `ResMut` instead
//~v ERROR: parameter takes `&mut ResMut` instead of a re-borrowed `ResMut`
fn mutable_reference_return<'a>(_res: &'a mut ResMut<Data>) -> usize {
    123
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return<'a>(res: &'a mut ResMut<Data>) -> &'a mut String {
    &mut res.0
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return_complex<'a>(
    res: &'a mut ResMut<Data>,
) -> Vec<(usize, &'a mut String)> {
    vec![(1, &mut res.0)]
}

fn main() {
    fn some_system(mut res: ResMut<Data>) {
        immutable_reference(&res);
        mutable_reference(&mut res);
        _ = mutable_reference_return(&mut res);
        _ = mutable_reference_bounded_return(&mut res);
        _ = mutable_reference_bounded_return_complex(&mut res);
    }

    App::new().add_systems(Update, some_system).run();
}
