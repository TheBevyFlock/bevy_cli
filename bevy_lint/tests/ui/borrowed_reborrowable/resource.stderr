error: parameter takes `&mut ResMut` instead of a re-borrowed `ResMut`
  --> tests/ui/borrowed_reborrowable/resource.rs:19:22
   |
19 | fn mutable_reference(_res: &mut ResMut<Data>) {
   |                      ^^^^^^^^^^^^^^^^^^^^^^^ help: use `ResMut` instead: `mut _res: bevy::prelude::ResMut<'_, Data>`
   |
note: the lint level is defined here
  --> tests/ui/borrowed_reborrowable/resource.rs:5:9
   |
5  | #![deny(bevy::borrowed_reborrowable)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: parameter takes `&mut ResMut` instead of a re-borrowed `ResMut`
  --> tests/ui/borrowed_reborrowable/resource.rs:25:33
   |
25 | fn mutable_reference_return<'a>(_res: &'a mut ResMut<Data>) -> usize {
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `ResMut` instead: `mut _res: bevy::prelude::ResMut<'_, Data>`

error: aborting due to 2 previous errors

